src/platform_impl/macos/mod.rs

```diff
@@ -1 +1,855 @@
+                    MenuItemType::Normal => Box::new(crate::MenuItem(MenuItem(c.clone()))),
+                    MenuItemType::Predefined => {
+                        Box::new(crate::PredefinedMenuItem(PredefinedMenuItem(c.clone())))
+                    }
+                    MenuItemType::Check => Box::new(crate::CheckMenuItem(CheckMenuItem(c.clone()))),
+                }
+            })
+            .collect()
+    }
+
+    pub fn init_for_nsapp(&self) {
+        unsafe { NSApp().setMainMenu_(self.ns_menu) }
+    }
+
+    pub fn remove_for_nsapp(&self) {
+        unsafe { NSApp().setMainMenu_(nil) }
+    }
+
+    pub fn show_context_menu_for_nsview(&self, view: id) {
+        unsafe {
+            let ns_menu_class = class!(NSMenu);
+            let ns_event: &mut Object = msg_send![NSApp(), currentEvent];
+            msg_send![ns_menu_class, popUpContextMenu: self.ns_menu withEvent: ns_event forView: view]
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct Submenu(Rc<RefCell<MenuChild>>);
+
+impl Submenu {
+    pub fn new(text: &str, enabled: bool) -> Self {
+        Self(Rc::new(RefCell::new(MenuChild {
+            type_: MenuItemType::Submenu,
+            text: strip_mnemonic(text),
+            enabled,
+            // submenu: Some(Menu::new()),
+            children: Some(Vec::new()),
+            ..Default::default()
+        })))
+    }
+
+    pub fn id(&self) -> u32 {
+        self.0.borrow().id()
+    }
+
+    pub fn make_ns_item_for_menu(&self, menu_id: u32) -> id {
+        let mut child = self.0.borrow_mut();
+        let ns_menu_item: *mut Object;
+        let ns_submenu: *mut Object;
+
+        unsafe {
+            ns_menu_item = NSMenuItem::alloc(nil).autorelease();
+            ns_submenu = NSMenu::alloc(nil).autorelease();
+
+            let title = NSString::alloc(nil).init_str(&child.text).autorelease();
+            let () = msg_send![ns_submenu, setTitle: title];
+            let () = msg_send![ns_menu_item, setTitle: title];
+            let () = msg_send![ns_menu_item, setSubmenu: ns_submenu];
+
+            if !child.enabled {
+                let () = msg_send![ns_menu_item, setEnabled: NO];
+            }
+        }
+
+        if let Some(children) = &child.children {
+            for item in children {
+                let item_type = &item.borrow().type_.clone();
+                let ns_item = match item_type {
+                    MenuItemType::Submenu => Submenu(item.clone()).make_ns_item_for_menu(menu_id),
+                    MenuItemType::Normal => MenuItem(item.clone()).make_ns_item_for_menu(menu_id),
+                    MenuItemType::Predefined => {
+                        PredefinedMenuItem(item.clone()).make_ns_item_for_menu(menu_id)
+                    }
+                    MenuItemType::Check => {
+                        CheckMenuItem(item.clone()).make_ns_item_for_menu(menu_id)
+                    }
```

Comment:
`menu_id` here is the parent menu, shouldn't you create a new `menu_id` for this newly created submenu and use that new id for the new children?